### PR TITLE
Camofy html image sources

### LINKED ISSUE #466: Camofy HTML content for static-pages
camofication of insecure resources currently only happens for markdown. We should have a look at whether this is also possible for HTML. Where did this come from?
A user found out it is possible to use HTML in static-page content, and subsequently tried to use images in this HTML. Sadly, this resulted in a failure to load images. We discovered that this is due to the image sources not being camofied, whereas with markdown content the image sources _are_ camofied.

So we should either not permit HTML content in static-pages, or we should also camofy HTML content.

### PR DESCRIPTION
**Pull request by DrumsnChocolate**
fix #466 
todo:
- [ ] test new stuff

### CODE REVIEW

File: app/helpers/markdown_helper.rb [hunk 1]
```diff
@@ -4,13 +4,41 @@ module MarkdownHelper
...
+    # warning: this regex may not be perfect. They rarely are.
+    # If you find an edge case, improve this regex!
+    # <img...something... src="url"...ending...
+    # or, the alternative quotes: <img...something... src='url'...ending...
+    # or, even without quotes: <img...something... src=url...ending...
+    # and the ...ending... can be either a space, a > or />
+    # note that we don't allow mismatched quotes like 'url" or shenanigans like that
+    # This regex contains two particularly useful features:
+    #  capturing groups, and lazy matching.
+    %r{<img([^>]*) src=(["']?)(.+?)\2( |>|/>)}
```
Comment (DrumsnChocolate) [inline app/helpers/markdown_helper.rb:42]: don't ask me why I had to put %r{} around it, I just followed rubocop's instructions.

File: app/helpers/markdown_helper.rb [hunk 2]
```diff
@@ -4,13 +4,41 @@ module MarkdownHelper
...
 
+  def sub_html(text)
+    text.gsub(html_img_regex) do
+      preceding_src = Regexp.last_match(1)
+      quote_mark = Regexp.last_match(2)
+      url = Regexp.last_match(3)
+      ending = Regexp.last_match(4)
+      "<img#{preceding_src} src=#{quote_mark}#{camo(url)}#{quote_mark}#{ending}"
+    end
+  end
```
Comment (DrumsnChocolate) [inline app/helpers/markdown_helper.rb:25]: I had to split things up into separate methods because of a linter limit on method line count

Comment (DrumsnChocolate) [reply app/helpers/markdown_helper.rb:25]: I think it's also more readable this way so I agree with the linter

File: app/helpers/markdown_helper.rb (hunk 1 above)
Comment (lodewiges) [reply app/helpers/markdown_helper.rb:42]: is there any documentaion online that i can use as a guide line for the review?

File: app/helpers/markdown_helper.rb (hunk 2 above)
Comment (lodewiges) [reply app/helpers/markdown_helper.rb:25]: makes 2

File: app/helpers/markdown_helper.rb [hunk 3]
```diff
@@ -4,13 +4,41 @@ module MarkdownHelper
   def camofy(markdown)
     return unless markdown
 
-    markdown.gsub(markdown_img_regex) do
+    markdown = sub_markdown(markdown)
```
Comment (lodewiges) [inline app/helpers/markdown_helper.rb:7]: why are you using the variable name markdown

File: app/helpers/markdown_helper.rb (hunk 1 above)
Comment (DrumsnChocolate) [reply app/helpers/markdown_helper.rb:42]: If you've never seen regexes before, they are a pain in the ass to review. They are kind of like a language in their own right, and the main way to understand regexes is to play around building them and testing them out on a site that shows you what the regex matches.
I've taken the regex here and prepared an example for you: https://regex101.com/r/RLd8UL/1
If you'd like to have a quick online meeting about how to understand this, let me know. I have time tomorrow

File: app/helpers/markdown_helper.rb (hunk 3 above)
Comment (DrumsnChocolate) [reply app/helpers/markdown_helper.rb:7]: I could change it to `text` if that scratches an itch for you. You're correct in suggesting this, I was just lazy

File: app/helpers/markdown_helper.rb (hunk 1 above)
Comment (lodewiges) [reply app/helpers/markdown_helper.rb:42]: I have seen regex before, and it is almost unreadable. thank you for the resource

File: app/helpers/markdown_helper.rb (hunk 3 above)
Comment (lodewiges) [reply app/helpers/markdown_helper.rb]: it is a small change and it improve readabilty

File: app/helpers/markdown_helper.rb (hunk 1 above)
Comment (DrumsnChocolate) [reply app/helpers/markdown_helper.rb:42]: The best way to test the regex is probably to come up with valid HTML img tags that are not matched by the regex. The edge cases I could think of have been captured in this one.

Comment (lodewiges) [reply app/helpers/markdown_helper.rb:42]: I am cheating a little bit by using chatgpt, it has some suggestions and explaination, i will check it and comment it

File: app/helpers/markdown_helper.rb [hunk 4]
```diff
@@ -1,16 +1,44 @@
...
+    # warning: this regex may not be perfect. They rarely are.
+    # If you find an edge case, improve this regex!
+    # <img...something... src="url"...ending...
+    # or, the alternative quotes: <img...something... src='url'...ending...
+    # or, even without quotes: <img...something... src=url...ending...
+    # and the ...ending... can be either a space, a > or />
+    # note that we don't allow mismatched quotes like 'url" or shenanigans like that
+    # This regex contains two particularly useful features:
+    #  capturing groups, and lazy matching.
+    %r{<img([^>]*) src=(["']?)(.+?)\2( |>|/>)}
```
Comment (lodewiges) [inline app/helpers/markdown_helper.rb:42]:
```suggestion
   %r{<img[^>]*\s+src=(["']?)([^'">]+)\1(?=[/>])}
```
this one also checks for space in between the elements and has better checking for the url

**Review by lodewiges — DISMISSED** [review]

**Comment by lodewiges** [comment]
Looks good, unable to test due to not having camo locally installed

**Comment by DrumsnChocolate** [comment]
> Looks good, unable to test due to not having camo locally installed

I don't have camo locally installed either. You don't need to. I believe it'll just use camo.csvalpha.nl when you're in development. Tests are slightly different, not sure how that's handled, but they should pass for you locally as well if you desire to run them.

File: app/helpers/markdown_helper.rb (hunk 4 above)
Comment (DrumsnChocolate) [reply app/helpers/markdown_helper.rb:42]: you seem to have removed the first capturing group. chatgpt has that tendency because it doesn't see you using its result anywhere in your regex. However, we definitely use this capturing group to produce the new text. What I do like, is the use of `\s`; that's any whitespace right?

What exactly does the latter part improve? Because I don't think it really improves anything. Let me break my thoughts down:
- **original**: `(.+?)\2( |>|/>)` where `\2` matches to the same value as the captured opening quote (either ', " or nothing at all). `(.+?)` _lazily_ matches any character. That means that it will begin with `src='` or `src="` or `src=` and then it will continue to match any character until the first time it encounters that same quotation mark again. Subsequently, after the quotation mark, it needs to match one of the three options in `( |>|/>)` which are a space, the `>` or `/>`.
- **your suggestion**: `([^'">]+)\1(?=[/>])}` where `\1` matches the same value as the captured opening quote. Note that `[^'">]` is unnecessarily restrictive: I don't know exactly the specification of HTML, but I can image that something like `src='lookatthisdoublequote"isntitamazing'` is valid. Notice the `"` in the middle? Your regex suggestion would stop the src value at that middle quote, because it does not pass the `[^'"]` check. And the latter part, `(?=[/>])` does not allow for any whitespace.


I do have some ideas based on your suggestion:
- use `\s` instead of spaces when I'm indicating whitespace
- I've reconsidered whether the last capturing group is necessary, but yes it is, because when we have `src=somesource`, we can only determine the end of the source value when we encounter a whitespace or / or >. But I can change it from `( |>|/>)` to `( |>|/)` which can be put simpler by writing `[ >/]`

Comment (DrumsnChocolate) [reply app/helpers/markdown_helper.rb]: I've applied the new ideas I mentioned in the latest commit

**Comment by lodewiges** [comment]
> > Looks good, unable to test due to not having camo locally installed
> 
> I don't have camo locally installed either. You don't need to. I believe it'll just use camo.csvalpha.nl when you're in development. Tests do things slightly different, not sure how that's handled, but they should pass for you locally as well if you desire to run them.

I will, i am having so trouble getting it to work

**Review by lodewiges — APPROVED** [review]
